core/services/request/files/FilesDataHandler.php

```diff
@@ -0,0 +1,237 @@
+     * @since $VID:$
+     * @param $files_data
+     * @return array
+     */
+    protected function fixFilesDataArray($files_data)
+    {
+        $sane_files_array = [];
+        foreach ($files_data as $top_key => $top_key_avlue) {
+            foreach ($top_key_avlue as $lower_key => $lower_key_value) {
+                foreach ($lower_key_value as $lowest_key => $lowest_key_value) {
+                    $next_data = [
+                        $top_key => [
+                            $lowest_key => $this->organizeFilesData($lowest_key_value, $lower_key, $lowest_key)
+                        ]
+                    ];
+                    $sane_files_array = array_merge_recursive(
+                        $sane_files_array,
+                        $next_data
+                    );
+                }
+            }
+        }
+        return $sane_files_array;
+    }
+
+    /**
+     * Recursively explores the array until it finds a leaf node, and tacks `$type` as a final index in front of it.
+     * @since $VID:$
+     * @param $data either 'name', 'tmp_name', 'size', or 'error'
+     * @param $type
+     * @return array
+     */
+    protected function organizeFilesData($data, $type)
+    {
+        $organized_data = [];
+        foreach ($data as $key => $val) {
+            if (is_array($val)) {
+                $organized_data[ $key ] = $this->organizeFilesData($val, $type);
+            } else {
+                $organized_data[ $key ][ $type ] = $val;
+            }
+        }
+        return $organized_data;
+    }
+
+    /**
+     * Takes the organized $_FILES array (where all file info is located at the same spot as you'd expect an input
+     * to be in $_GET or $_POST, with all the file's data located side-by-side in an array) and creates a
+     * multi-dimensional array of FileSubmissionInterface objects.
+     * @since $VID:$
+     * @param $organized_files
+     * @return array
+     * @throws UnexpectedValueException
+     * @throws InvalidArgumentException
+     */
+    protected function createFileObjects($organized_files)
+    {
+        if (!is_array($organized_files)) {
+            throw new UnexpectedValueException(
+                sprintf(
+                    esc_html__(
+                        'Unexpected PHP $organized_files data format. "%1$s" was expected to be an array.',
+                        'event_espresso'
+                    ),
+                    (string) $organized_files
+                )
+            );
+        }
+        $objs = [];
+        foreach ($organized_files as $key => $value) {
+            if (isset($value['name'], $value['tmp_name'], $value['size'])) {
+                $objs[ $key ] = new FileSubmission(
+                    $value['name'],
+                    $value['tmp_name'],
+                    $value['size'],
+                    $value['error']
+                );
+            } else {
+                $objs[ $key ] = $this->createFileObjects($value);
```

Comment:
if something goes wrong with the initial form data, then this could throw an unintended exception, if `$value` is not another array, no?

Reply:
yeah, although with refactoring to use a collection this line got removed